libraries/AP_GPS/AP_GPS_MAV.cpp

```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
 // corresponding gps data appropriately;
 void AP_GPS_MAV::handle_msg(const mavlink_message_t *msg)
 {
-    mavlink_gps_input_t packet;
-    mavlink_msg_gps_input_decode(msg, &packet);
-
-    bool have_alt    = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_ALT) == 0);
-    bool have_hdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HDOP) == 0);
-    bool have_vdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VDOP) == 0);
-    bool have_vel_h  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_HORIZ) == 0);
-    bool have_vel_v  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_VERT) == 0);
-    bool have_sa     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_SPEED_ACCURACY) == 0);
-    bool have_ha     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HORIZONTAL_ACCURACY) == 0);
-    bool have_va     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VERTICAL_ACCURACY) == 0);
-
-    state.time_week     = packet.time_week;
-    state.time_week_ms  = packet.time_week_ms;
-    state.status = (AP_GPS::GPS_Status)packet.fix_type;
-
-    Location loc = {};
-    loc.lat = packet.lat;
-    loc.lng = packet.lon;
-    if (have_alt) {
-        loc.alt = packet.alt;
+    switch (msg->msgid) {
+
+        case MAVLINK_MSG_ID_GPS_INPUT: {
+            mavlink_gps_input_t packet;
+            mavlink_msg_gps_input_decode(msg, &packet);
+
+            bool have_alt    = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_ALT) == 0);
+            bool have_hdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HDOP) == 0);
+            bool have_vdop   = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VDOP) == 0);
+            bool have_vel_h  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_HORIZ) == 0);
+            bool have_vel_v  = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VEL_VERT) == 0);
+            bool have_sa     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_SPEED_ACCURACY) == 0);
+            bool have_ha     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_HORIZONTAL_ACCURACY) == 0);
+            bool have_va     = ((packet.ignore_flags & GPS_INPUT_IGNORE_FLAG_VERTICAL_ACCURACY) == 0);
+
+            state.time_week     = packet.time_week;
+            state.time_week_ms  = packet.time_week_ms;
+            state.status = (AP_GPS::GPS_Status)packet.fix_type;
+
+            Location loc = {};
+            loc.lat = packet.lat;
+            loc.lng = packet.lon;
+            if (have_alt) {
+                loc.alt = packet.alt;
```

Comment:
I know this is not new but the packet.alt is in meters, and you are storing it to a Location struct, which isn't correct. (Either that or I'm way to tired). I think this one needs to be converted into cm's before storing. (Either that or the other one needs to come out of cm's).